conda/__init__.py

```diff
@@ -148,3 +139,209 @@ def _default(self, obj):
 
 _default.default = JSONEncoder().default
 JSONEncoder.default = _default
+
+
+# inspired by deprecation (https://deprecation.readthedocs.io/en/latest/) and
+# CPython's warnings._deprecated
+class _deprecated:
+    _version: tuple[int, ...]
+    _category: Warning
+    _message: Callable[str, str]
+    _argument: str | None = None
+    _rename: str | None = None
+    _stack: int
+
+    @staticmethod
+    def _parse_version(version: str) -> tuple[int, ...]:
+        """A naive version parser to avoid circular imports. Do not use for other purposes."""
+
+        def try_int(string: str) -> int | str:
+            try:
+                return int(string)
+            except ValueError:
+                return string
+
+        return tuple(map(try_int, version.split(".")))
+
+    def __init__(
+        self,
+        deprecate_in: str,
+        remove_in: str,
+        *,
+        addendum: str | None = None,
+        stack: int = 0,
+    ):
+        """Deprecation decorator for functions, methods, & classes.
+
+        Args:
+            deprecate_in: Version in which code will be marked as deprecated.
+            remove_in: Version in which code is expected to be removed.
+            addendum: Optional additional messaging. Useful to indicate what to do instead.
+            stack: Optional stacklevel increment.
+        """
+        deprecate_version = self._parse_version(deprecate_in)
+        remove_version = self._parse_version(remove_in)
+
+        addendum = f" {addendum}" if addendum else ""
+        if self._version < deprecate_version:
+            self._category = PendingDeprecationWarning
+            message = f"{{name}} is pending deprecation and will be removed in {remove_in}."
+        elif self._version < remove_version:
+            self._category = DeprecationWarning
+            message = f"{{name}} is deprecated and will be removed in {remove_in}."
+        else:
+            self._category = None
+            message = f"{{name}} was slated for removal in {remove_in}."
+        self._message = lambda name: f"{message}{addendum}".format(name=name)
+
+        self._stack = stack
+
+    def __call__(self, func: Callable) -> Callable:
+        """Deprecation decorator for functions, methods, & classes."""
+        # detect function name
+        fullname = f"{func.__module__}.{func.__qualname__}"
+        if self._argument:
+            fullname = f"{func.__module__}.{func.__qualname__}({self._argument})"
+
+        # alert developer that it's time to remove something
+        if not self._category:
+            raise RuntimeError(self._message(fullname))
```

Comment:
It might be useful to configure the type of exception that is used as a fallback.

Reply:
So a new `DeprecationError` derived from `RuntimeError`?

Reply:
Yeah, or even a parameter. But an own exception class would suffice as well.